week1/intro.cpp

```diff
@@ -0,0 +1,34 @@
+class Foo {
+public:
+    mutex m;
+    condition_variable cv;
+    int turn = 1;
+    
+    Foo() {
+    }
+
+    void printn(function<void()>& printFunc, int n) {
+        {
+            unique_lock lk(m);
+            cv.wait(lk, [this,n](){return turn == n;});
+        }
+        printFunc();
+        {
+            unique_lock lk(m);
+            turn++;
+        }
+        cv.notify_all();
+    }
```

Comment:
proof of correctness: 

In all cases in the code we always release lock immediately after acquiring it. Therefore we can always assume that acquiring a lock in any part of the code will never fail. 

- If is not the threads current turn upon entering the first block then it will stay on the conditional wait. 
- If it is the threads current turn then we will fall through the conditional wait. Since all other threads will stall on the condition, only one thread can print a time (and only when it is it's the correct turn).

As prior reasoning, we will always acquire the lock in the second block and be able to increment the turn after we finish printing. Since we always notify after finishing, we guarantee that the thread with the next turn will never get stalled on the condition. Therefore by induction we will never stall.

Reply:
small
> release lock immediately
unclear what "immediatelly" means and its significance to the proof,   I can guess you imply that we don't invoke any other code with unknown/unbounded behavior there?
> acquiring a lock in any part...
perhaps better phrased - code fragments between lock and unlock always complete successfully, and therefore you conclude that lock can't fail.

Although I don't require this for proofs in this course - if you are thinking about failure edge cases - one further thing  to think about - in Java (using Thread.interrupt()) and in native code using signals - wait() can be interrupted.